### ISSUE TITLE
rowData Clustering

### ISSUE THREAD

**Issue by ChouaibB** [microbiome/mia#186]
Hi,

Adding a function for clustering rowData (of a TreeSe object, e.g. taxonomic, metabolomic.. data) could be useful in analysis. For instance, performing hierarchical clustering on rowData based on correlation, then summarizing clusters by their mean abundance or sum. This could simplify analysis as it reduces the multiple statistical testing in many occasions. The following are examples of co-abundance groups "CAGs" using clustering methods for analysis:
https://www.nature.com/articles/nature11319
https://pubmed.ncbi.nlm.nih.gov/31370880/

**Comment by antagomir** [comment]
I think this would be a useful addition to the exploration toolkit but prioritization w.r.t other issues can be discussed. This can be done also without wrappers, the main purpose of a wrapper would be to make it faster and more standardized to run such analyses. Experienced users can then build their own custom workflows always, if needed.

**Comment by ChouaibB** [comment]
Hi,
I wanted to discuss and ask your opinions about implementation.
I tried to think how the clustering data/information could be used by the end-user.
For instance:
* rowClust: would just perform the clustering, then create a new column at colData with cluster ids. This way the user could still use original data (not aggregated) to perhaps e.g. perform PCA with original data and color with cluster ids, or subset by cluster ids and perform analysis separately... So as not to have extra storage of the already existing original data plus the newly aggregated cluster data.
* aggregateByCluster: (using the previous function) in this case the data would be aggregated(sum/mean..) and returned as a new TreeSe object, with corresponding other parts (colData..) of the original TreeSe (similarly to the aggregateByRank fashion). Then the user can keep on going with the newly aggregated data and delete the original one; and freeing up memory, if the user wants, especially with large datasets.

Optionally:
* perhaps at miaViz some visualization of each cluster and which taxa (or metabolites) grouped in it? (e.g. dendogram with rectangles when using hclust, or just a pie chart with each section as a cluster showing labels of original taxa when using kmeans)
* Are clustering diagnostics needed? perhaps a verbose arg at rowClust printing the summary of clustering? (e.g. number of clusters after choosing a cut-off for hclust or cluster centroids for kmeans)

Of course the visualization could also be just referred to at vignette/OMA with examples using other packages; not to reinvent the wheel.

What do you think? 
Thanks

**Comment by antagomir** [comment]
It would be great if you can give it a try. 

We would need:

1. Function that determines CAGs; this could be named `rowClust` because this is descriptive. Perhaps an alias for `CAG` could be helpful as it connects with the better known concept and some users might prefer that. We could provide the options that are available through *hclust* or *fastcluster* (faster one, hence preferable?) and provide recommended defaults. It will be helpful to browse through the above mentioned references and perhaps some others to compare the defaults to the common use cases in the literature. This function will add the CAG information to `rowData` (not `colData`!) since the grouping is for taxa (rows) and not for samples (cols).
2. Function that collapses data to CAGs. Primarily, we should check if existing function (e.g. `agglomerateByRank`) could be used for this and if it should be renamed or aliased (e.g. `aggregateByRowData`) to extend the scope to other `rowData` fields besides taxa. This aggregation step should be a separate function that can operate on the output from `rowClust`. 

After implementing 1-2, we could consider providing an additional wrapper that combines 1-2 and could determine the clusters and collapse the data at one go. 

To summarize the detected CAGs, it might be enough to use base functionality. Start with a simple example in *OMA* (e.g. number of clusters and histogram of their sizes). We can discover later if dedicated new functions are really needed for this. Better to now overdo. Same for visualization, I would not add new visualization tools for CAGs in the first phase. If clear needs arise, this can always be reconsidered.

**Comment by ChouaibB** [comment]
Thank you, I will give it a try.

**Comment by microsud** [comment]
I recently applied this concept for decontaminating low biomass nasopharyngeal samples. Below are the codes. Might be helpful for a more general implementation.
https://github.com/RIVM-IIV-Microbiome/ILI-Respiratory-Microbiota-2021/blob/9261c8d1e607e85293aa66a25be319ba7ec6de07/02_DecontaminateMicrobiotaData.Rmd#L1012

**Comment by FelixErnst** [comment]
I think you need to think about this a bit more and focus on the actual steps of an analysis. (In any event `rowClust` is not a good name for any function, since the same clustering method would and should also be available for cols)

I can already see that any type of clustering would start with generating a factor along rows. And rows of the same cluster should be merged. Is this assumption correct?

If yes, why not use `mergeRows`, an already implemented function, which requires just the clusting factor? Any additional diagnostic can be derived from this factor alone not resorting to printing out some summaries.

**Comment by antagomir** [comment]
Yes. 

`mergeRows` has been used in the PR.

Agree about rows/cols, although the initial motivation was to have a wrapper for Co-Abundant (taxonomic) Groups (CAGs), which is a well-known concept.

**Comment by FelixErnst** [comment]
OK, but than add a function to add the CAG to colData. They can be used as factors for any number of purposes, but this doesn't require a function dedicated only for one task

**Comment by antagomir** [comment]
In fact this is what the current rowClust function does. We can change the name to getCAGs but motivation for the rowClust name was that it is a bit more descriptive. And the clustering method does not matter, we could support any number of clustering methods that can be easily incorporated to rowClust. If this is not feasible, though, then perhaps a more specific function, like getCAGs, with a more narrowly defined range of choices would be the way to go.

**Comment by FelixErnst** [comment]
IMO the flexibility shown in the OSCA book regarding clustering should be carefully examined. A wrapper to solve all clustering problem is be definition not easy compared to implementing the clustering by itself. Thats probagly why nobody implemented it, yet.

I have mentioned my issue with the name `rowClust`.

**Comment by antagomir** [comment]
I agree that a new package might be useful, we aim to activate with that soon. This is a relevant CRAN package for phylogeny-aware dissimilarity measures [GUniFrac](https://cran.r-project.org/web/packages/GUniFrac/index.html). This issue can be closed after we have initiated that new cluster-oriented pkg.

**Comment by antagomir** [comment]
@BananaCancer could you check if all relevant aspects from here have been dealt with? 

If yes, then close the issue.
